pkg/urbit/vere/io/term.c

```diff
@@ -1036,6 +1042,169 @@ u3_term_ef_ctlc(void)
   _term_it_refresh_line(uty_u);
 }
 
+/* _term_it_put_color(): put ansi color id on lin_w. RETAINS col.
+*/
+static void
+_term_it_put_color(c3_w* lin_w,
+                   u3_noun col)
+{
+  switch ( col ) {
+    default:
+    case u3_nul: *lin_w = '9'; break;
+    case 'k':    *lin_w = '0'; break;
+    case 'r':    *lin_w = '1'; break;
+    case 'g':    *lin_w = '2'; break;
+    case 'y':    *lin_w = '3'; break;
+    case 'b':    *lin_w = '4'; break;
+    case 'm':    *lin_w = '5'; break;
+    case 'c':    *lin_w = '6'; break;
+    case 'w':    *lin_w = '7'; break;
+  }
+}
+
+/* _term_it_put_decoration(): put ansi sgr code on lin_w. RETAINS dec.
+*/
+static void
+_term_it_put_decoration(c3_w* lin_w,
+                        u3_noun dec)
+{
+  switch ( dec ) {
+    default:
+    case u3_nul: *lin_w = '0'; break;
+    case c3__br: *lin_w = '1'; break;
+    case c3__un: *lin_w = '4'; break;
+    case c3__bl: *lin_w = '5'; break;
+  }
+}
+
+/* _term_it_show_stub(): send styled text to terminal as ansi escape sequences
+*/
+static void
+_term_it_show_stub(u3_utty* uty_u,
+                   u3_noun tub)
+{
+  c3_w tuc_w = u3qb_lent(tub);
+
+  //  count the amount of characters across all stubs
+  //
+  c3_w lec_w = 0;
+  {
+    u3_noun nub = tub;
+    while ( u3_nul != nub ) {
+      u3_noun nib = u3t(u3h(nub));
+      lec_w = lec_w + u3qb_lent(nib);
+      nub = u3t(nub);
+    }
+  }
+
+  //  allocate enough memory for every display character, plus styles
+  //
+  //NOTE  we use max 20 characters per styl for escape codes:
+  //      3 for opening, 4 for decorations, 4 for colors, 4 for closing,
+  //      and 5 as separators between decorations and colors.
+  //
+  c3_w* lin_w = c3_malloc( 4 * (lec_w + (20 * tuc_w)) );
+
+  //  write the contents to the buffer,
+  //  tracking total and escape characters written
+  //
+  c3_w   i_w = 0;
+  c3_w sap_w = 0;
+  {
+    u3_noun nub = tub;
+    while ( u3_nul != nub ) {
+      u3_noun nib   = u3t(u3h(nub));
+      u3_noun tyl   = u3h(u3h(nub));
+      u3_noun dec   = u3h(tyl);
+      u3_noun bag   = u3h(u3t(tyl));
+      u3_noun fog   = u3t(u3t(tyl));
+      c3_o    tyl_o = c3n;
```

Comment:
I think this would be clearer as

```
u3_noun tyl, nib, dec, bag, fog;
u3x_cell(u3h(nub), &tyl, &nib);
u3x_trel(tyl, &dec, &bag, &fog);
```

And the horizontal spacing can be reduced.